dev-guide/feature-gate.md

```diff
@@ -0,0 +1,293 @@
+}
+
+type FeatureGateDetails struct {
+	// version matches the version provided by the ClusterVersion and in the ClusterOperator.Status.Versions field.
+	// +kubebuilder:validation:Required
+	// +required
+	Version string `json:"version"`
+	// enabled is a list of all feature gates that are enabled in the cluster for the named version
+	// +optional
+	Enabled []FeatureGateAttributes `json:"enabled"`
+	// disabled is a list of all feature gates that are disabled in the cluster for the named version
+	// +optional
+	Disabled []FeatureGateAttributes `json:"disabled"`
+}
+
+type FeatureGateAttributes struct {
+	// name is the name of the FeatureGate
+	// +kubebuilder:validation:Pattern=`^([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$`
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+
+	// possible (probable?) future additions include
+	// 1. support level (Stable, ServiceDeliveryOnly, TechPreview, DevPreview)
+	// 2. description
+}
+```
+
+#### openshift/library-go
+The existing config observer will be updated to be a drop-in replacement (this covers a dozen or so operators).
+Operators that don't use that pattern can use one like this
+
+```go
+featureGateAccessor := NewFeatureGateAccess(args)
+go featureGateAccessor.Run(ctx)
+
+// wait for feature gates to observed.
+select{
+case <- featureGateAccessor.InitialFeatureGatesObserved():
+	enabled, disabled, _ := featureGateAccessor.CurrentFeatureGates()
+	klog.Infof("FeatureGates initialized: enabled=%v  disabled=%v", enabled, disabled)
+case <- time.After(1*time.Minute):
+	klog.Errorf("timed out waiting for FeatureGate detection")
+	return fmt.Errorf("timed out waiting for FeatureGate detection")
+}
+
+// continue from here.
+// also open to changes to wire up to k/k FeatureGates if there is demand.
+```
+
+library-go also has code allow a "default" payload version (one that hasn't been substituted) to mean, "give me the 
+feature gates for the latest CVO version".
+This will support the development patch use-cases.
+
+### Workflow Description
+
+#### Variation [optional]
+
+### API Extensions
+
+No admission webhooks are needed.
+If complicated validation is required, there is already a FeatureGate admission plugin hardcoded in our openshift/kubernetes
+fork and we'll use that.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+### Drawbacks
+
+## Design Details
+
+### Open Questions [optional]
+
+### Test Plan
+
+Unit tests are already added to implementation PRs.
+Clusters will not install without the controller functional and we don't support changing feature gates back from
+TechPreviewNoUpgrade, so e2e tests are impractical.
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+Since this is actually controlling the gates, it is not practical to pre-test it.
+Once unit tests pass and clusters successfully install equivalently using this mechanism, PRs will be opened against 
```

Comment:
I am interested how the MCO and conversely the Kubelet would digest the configured FeatureGates at bootstrap.

MCO Reference for Features Gates: https://github.com/openshift/machine-config-operator/pull/2668